File: core/src/com/alexpi/awesometanks/entities/Tank.java
package com.alexpi.awesometanks.entities;

import com.badlogic.gdx.Preferences;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.Fixture;
import com.badlogic.gdx.physics.box2d.FixtureDef;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.World;
import com.alexpi.awesometanks.utils.Constants;
import com.alexpi.awesometanks.weapons.Weapon;
import com.badlogic.gdx.scenes.scene2d.Group;


/**
 * Created by Alex on 02/01/2016.
 */
public class Tank extends DamageableActor{

    private final Sprite bodySprite, wheelsSprite;
    private int currentWeapon;
    private Weapon[] weapons;
    private Body body;
    private Fixture fixture;
    public float jX, jY, size, currentAngleRotation, desiredAngleRotation, rotationSpeed, movementSpeed;
    public boolean alive, isMoving, hasToShoot, allowSounds;
    public Vector2 movement;
    public int money;
    public float visibilityRadius;
    private AssetManager manager;
    private Group entityGroup;

    public Tank(AssetManager manager, Group entityGroup, World world, Vector2 position, Preferences gameValues, Color color, DamageListener listener, int money, boolean sounds){
        super(manager, listener,1200 + gameValues.getInteger("health")*200 );
        allowSounds = sounds;this.manager = manager;this.money = money;
        bodySprite = new Sprite(manager.get("sprites/tank_body.png",Texture.class));
        wheelsSprite = new Sprite(manager.get("sprites/tank_wheels.png",Texture.class));
        this.entityGroup = entityGroup;
        size = .75f;
        currentAngleRotation = desiredAngleRotation = jX = jY = 0;
        alive = true;isMoving=false;
        movement = new Vector2();

        weapons = new Weapon[7];

        for(int i = 0; i < 7; i++)
            weapons[i] = Weapon.getWeaponAt(i, manager, gameValues.getInteger("ammo" + i), gameValues.getInteger("power" + i), Constants.TANK_BULLET_MASK, sounds);

        visibilityRadius = 1.5f+ gameValues.getInteger("visibility")/2f;
        rotationSpeed = .07f+gameValues.getInteger("rotation") /40f;
        movementSpeed = 130f + gameValues.getInteger("speed")*10f;
        BodyDef bodyDef = new BodyDef();
        bodyDef.type = BodyDef.BodyType.DynamicBody;
        bodyDef.position.set(position.x + .5f, position.y + .5f);

        PolygonShape shape = new PolygonShape();
        shape.setAsBox(size / 2, size / 2);

        FixtureDef fixtureDef = new FixtureDef();
        fixtureDef.density = 10f;
        fixtureDef.shape = shape;
        fixtureDef.filter.categoryBits = Constants.CAT_TANK;
        fixtureDef.filter.maskBits = Constants.CAT_BLOCK | Constants.CAT_ITEM | Constants.CAT_ENEMY | Constants.CAT_BULLET;

        body = world.createBody(bodyDef);
        fixture = body.createFixture(fixtureDef);
        fixture.setUserData(this);
        body.setUserData(this);

        shape.dispose();

        setSize(size * Constants.TILE_SIZE, size * Constants.TILE_SIZE);
        setOrigin(getWidth() / 2, getHeight() / 2);
        setPosition((body.getPosition().x - size / 2) * getWidth(), (body.getPosition().y - size / 2) * getHeight());
        setColor(color);
        setJustHit(true);
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        getColor().a*=parentAlpha;
        batch.draw(wheelsSprite, getX(), getY(), getOriginX(), getOriginY(), getWidth(), getHeight(), getScaleX(), getScaleY(), getRotation());
        batch.setColor(getColor());
        batch.draw(bodySprite, getX(), getY(), getOriginX(), getOriginY(), getWidth(), getHeight(), getScaleX(), getScaleY(), getRotation());
        batch.draw(weapons[currentWeapon].sprite, getX(), getY(), getOriginX(), getOriginY(), getWidth(), getHeight(), getScaleX(), getScaleY(), weapons[currentWeapon].sprite.getRotation());
        batch.setColor(1, 1, 1, parentAlpha);
        super.draw(batch, parentAlpha);
    }

    @Override
    public void act(float delta) {
        if(isMoving){
            movement.set(movementSpeed * jX * delta, movementSpeed * jY * delta);
            setDesiredAngleRotation(jX,jY);
            updateAngleRotation();
            body.setLinearVelocity(movement);
            body.setTransform(body.getPosition(), currentAngleRotation);
        }else{
            body.setLinearVelocity(0, 0);
            body.setAngularVelocity(0);
        }
        getCurrentWeapon().updateAngleRotation(rotationSpeed);

        if(hasToShoot && getCurrentWeapon().hasRotated() && isAlive()){
            getCurrentWeapon().shoot(manager, entityGroup, body.getWorld(), body.getPosition(), damageListener);
            //hasToShoot = false;
        }
        setPosition((body.getPosition().x - size / 2) * Constants.TILE_SIZE, (body.getPosition().y - size / 2) * Constants.TILE_SIZE);
        setRotation(body.getAngle() * MathUtils.radiansToDegrees);
        super.act(delta);
    }
    public void setDirection(float x, float y){jX = x;jY = y;}
    public void setDesiredAngleRotation(float x, float y) {
        desiredAngleRotation = (float) Math.atan2(y,x);
        if(desiredAngleRotation < 0)
            desiredAngleRotation +=Math.PI*2;
    }
    public void updateAngleRotation() {
        float diff = desiredAngleRotation - currentAngleRotation;
        if(diff<0)diff+=Math.PI*2;

        if(diff>=Math.PI){currentAngleRotation-=.1f;diff-=Math.PI;}
        else if(diff<Math.PI) currentAngleRotation+=.1f;

        if(diff<.1f)currentAngleRotation = desiredAngleRotation;

        if(currentAngleRotation>Math.PI*2)currentAngleRotation= 0;
        else if(currentAngleRotation<0)currentAngleRotation= (float) (Math.PI*2);
    }

    @Override
    public void detach(){
        super.detach();
        body.destroyFixture(fixture);
        body.getWorld().destroyBody(body);
        remove();
    }

    public void saveProgress(Preferences gameValues){
        for(int i =0; i<weapons.length;i++)
            gameValues.putInteger("ammo" + i, weapons[i].getAmmo());}

    public void setCurrentWeapon(int num){
        currentWeapon = num;
    }
    public Weapon getCurrentWeapon(){return weapons[currentWeapon];}
    public float getPosY() {return body.getPosition().y;}
    public float getPosX() {return body.getPosition().x;}
    public Body getBody() {return body;}
}

File: core/src/com/alexpi/awesometanks/entities/tank/Tank.kt
package com.alexpi.awesometanks.entities.tank

import com.alexpi.awesometanks.entities.DamageableActor
import com.alexpi.awesometanks.utils.Constants
import com.alexpi.awesometanks.weapons.Weapon
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.*
import com.badlogic.gdx.scenes.scene2d.Group
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Created by Alex on 02/01/2016.
 */
abstract class Tank(
    protected val manager: AssetManager,
    protected val entityGroup: Group,
    world: World,
    position: Vector2,
    private val bodySize: Float,
    private val rotationSpeed: Float,
    private val movementSpeed: Float,
    categoryBits: Short,
    maskBits: Short,
    maxHealth: Float,
    isFreezable: Boolean,
    protected val allowSounds: Boolean,
    ) : DamageableActor(manager, maxHealth, true, isFreezable){

    private val bodySprite: Sprite = Sprite(manager.get("sprites/tank_body.png", Texture::class.java))
    private val wheelsSprite: Sprite = Sprite(manager.get("sprites/tank_wheels.png", Texture::class.java))

    val body: Body
    private val fixture: Fixture
    var currentAngleRotation: Float = 0f
    private var desiredAngleRotation: Float = 0f
    set(value) {
        field = if (value < -MathUtils.PI) value + MathUtils.PI
        else if(value > MathUtils.PI) value - MathUtils.PI
        else value
    }
    var isMoving: Boolean = false
    var isShooting = false
    private val movement: Vector2 = Vector2()
    override fun draw(batch: Batch, parentAlpha: Float) {
        color.a *= parentAlpha
        batch.draw(wheelsSprite, x, y, originX, originY, width, height, scaleX, scaleY, rotation)
        batch.color = color
        batch.draw(bodySprite, x, y, originX, originY, width, height, scaleX, scaleY, rotation)
        batch.draw(
            getCurrentWeapon().sprite,
            x,
            y,
            originX,
            originY,
            width,
            height,
            scaleX,
            scaleY,
            getCurrentWeapon().sprite.rotation
        )
        batch.setColor(1f, 1f, 1f, parentAlpha)
        super.draw(batch, parentAlpha)
    }

    override fun act(delta: Float) {
        super.act(delta)
        if (isMoving) {
            updateAngleRotation()
            body.setLinearVelocity(movement.x * delta, movement.y * delta)
            body.setTransform(body.position, currentAngleRotation)
        } else {
            body.setLinearVelocity(0f, 0f)
            body.angularVelocity = 0f
        }
        getCurrentWeapon().updateAngleRotation(rotationSpeed)
        if (isShooting && getCurrentWeapon().hasRotated() && isAlive) {
            getCurrentWeapon().shoot(
                manager,
                entityGroup,
                body.world,
                body.position
            )
        }
        setPosition(
            (body.position.x - bodySize*.5f) * Constants.TILE_SIZE,
            (body.position.y - bodySize*.5f) * Constants.TILE_SIZE
        )
        rotation = body.angle * MathUtils.radiansToDegrees
    }

    //x and y values must be normalized
    fun setOrientation(x: Float, y: Float) {
        desiredAngleRotation = MathUtils.atan2(y, x)
        movement.set(x * movementSpeed, y * movementSpeed)
    }

    //In radians
    fun setOrientation(rotationAngle: Float) {
        desiredAngleRotation = rotationAngle
        movement.set(MathUtils.cos(rotationAngle) * movementSpeed, MathUtils.sin(rotationAngle) * movementSpeed)
    }

    private fun updateAngleRotation() {
        val change = .1f
        var diff = desiredAngleRotation - currentAngleRotation
        if (diff < -MathUtils.PI) diff += MathUtils.PI * 2
        else if (diff > MathUtils.PI) diff -= MathUtils.PI * 2

        if(diff == 0f) return
        else if (abs(diff) < change) {
            currentAngleRotation = desiredAngleRotation
            return
        }

        currentAngleRotation += min(change, max(-change, diff))
    }

    override fun detach() {
        body.destroyFixture(fixture)
        body.world.destroyBody(body)
        remove()
    }

    abstract fun getCurrentWeapon(): Weapon

    val posY: Float
        get() = body.position.y
    val posX: Float
        get() = body.position.x

    init {
        val bodyDef = BodyDef()
        bodyDef.type = BodyDef.BodyType.DynamicBody
        bodyDef.position.set(position.x + .5f, position.y + .5f)
        val shape = PolygonShape()
        shape.setAsBox(bodySize / 2, bodySize / 2)
        val fixtureDef = FixtureDef()
        fixtureDef.density = 10f
        fixtureDef.shape = shape
        fixtureDef.filter.categoryBits = categoryBits
        fixtureDef.filter.maskBits = maskBits
        body = world.createBody(bodyDef)
        fixture = body.createFixture(fixtureDef)
        fixture.userData = this
        body.userData = this
        shape.dispose()
        setSize(bodySize * Constants.TILE_SIZE, bodySize * Constants.TILE_SIZE)
        setOrigin(width / 2, height / 2)
        setPosition((body.position.x - bodySize / 2) * width, (body.position.y - bodySize / 2) * height)
    }
}